### PR TITLE
clientv3: fix health balancer, retry

### PR DESCRIPTION
**Pull request by gyuho**
Fix most of bugs in balancer with stale endpoints.
Fix sync with stale upc channel in retry path.
Will send a separate PR to remove redundant retry code.

### CODE REVIEW

**Comment by xiang90** [comment]
test is still failing, but only one it seems.

File: clientv3/balancer.go [hunk 1]
```diff
@@ -45,7 +45,9 @@ type balancer interface {
 	// pinned returns the current pinned endpoint.
 	pinned() string
 	// endpointError handles error from server-side.
-	endpointError(addr string, err error)
+	// returns true when addr is marked as unhealthy.
+	// returns false when addr was stale endpoint.
```
Comment (xiang90) [inline clientv3/balancer.go]: what is a stale endpoint? this func should do only one thing as reporting endpoint error, not checking if the endpoint is stale.

File: clientv3/retry.go [hunk 2]
```diff
@@ -23,65 +23,96 @@ import (
...
 			}
-			// mark this before endpoint switch is triggered
-			c.balancer.endpointError(pinned, err)
-			notify := c.balancer.ConnectNotify()
-			if s, ok := status.FromError(err); ok && s.Code() == codes.Unavailable {
-				c.balancer.next()
+
+			// if temporary gRPC connection error, then endpoint switch but no retry
+			// e.g. transport.ErrConnClosing when server closed transport, failing node
+			if tv, ok := err.(transport.ConnectionError); ok && tv.Temporary() {
```
Comment (xiang90) [inline clientv3/retry.go]: if it is temporary, we should retry on the same endpoint not switch to another one.

Comment (xiang90) [reply clientv3/retry.go]: if it is not, we should retry on another endpoint if possible. also it is not clear to me if gRPC will ever return connection error. can you verify it?

File: clientv3/retry.go [hunk 3]
```diff
@@ -23,65 +23,96 @@ import (
...
+				s.Code() == codes.Internal ||
+				s.Code() == codes.DeadlineExceeded {
+				exist := c.balancer.endpointError(pinned, err)
+				if exist || s.Message() == "there is no address available" {
+					c.balancer.next()
+				}
+				// only retry immutable RPCs ("put at most once semantics")
+				// or mutable RPCs with "there is no address available"
+				if !isWrite || s.Message() == "there is no address available" {
+					select {
```
Comment (xiang90) [inline clientv3/retry.go]: we really should not duplicate this code three times in one func.

Comment (gyuho) [reply clientv3/retry.go]: Agree, will fix/clean-up tomorrow.

**Comment by gyuho** [comment]
GitHub sync doesn't seem to work... Will create a new PR after investigating test failures.

**Pull request closed without merging**